src/main/scala/uclid/ConcreteSimulator.scala

```
@@ -0,0 +1,1293 @@
+                        if (!terminate) {
+                            concreteContext.assignUndefVar(module.inputs,true)
+                            simulate_stmt(concreteContext, next.body,a,module)
+
+                            checkProperties(properties,concreteContext,a,module)
+                            trace(a) = concreteContext.cloneObject;
+                            terminateInt = a;   
+                        } 
+                        else {
+                            if (!terminate_printed) {
+                                printDebug(s"Failed on iteration ${a-1}")
+                                terminate_printed = true
+                            }    
+                        }
+                    }
+                }
+                case _ => {}
+            }
+        }
+        
+        if(needToPrintResults){
+            UclidMain.printResult("%d assertions passed.".format(passCount))
+            UclidMain.printResult("%d assertions failed.".format(failCount))
+            UclidMain.printResult("%d assertions indeterminate.".format(undetCount))
+
+            if(needToPrintTrace){
+                printConcretetTrace(trace, printTraceCmd.args, printTraceCmd.argObj)
+            }
+        }
+        // print(proofResults.toList)
+        return proofResults.toList}
+
+
+    def simulate_stmt (context: ConcreteContext, stmt: Statement, iter: Int, module: Module): Unit = {
+        printDebug("Simulate Stmt: "+stmt.toString)
+        stmt match {
+            case AssignStmt(lhss, rhss) => {
+            
+                //println("Simulate assign Stmt: "+stmt.toString)
+                val rhseval = rhss.map(rhs => evaluate_expr(context, rhs))
+                for((lhssid,i)<-lhss.view.zipWithIndex){
+                    if(rhseval(i).isInstanceOf[ConcreteUndef]){
+                        printDebug("We hit a undefine value when assigning "+rhseval(i).toString)
+                        //Leiqi:
```

this seems like it should throw an error? What is a concreteundef?